Packages/com.unity.inputsystem/InputSystem/InputManagerStateMonitors.cs

```diff
@@ -351,6 +351,11 @@ internal unsafe void FireStateChangeNotifications(int deviceIndex, double intern
         {
             Debug.Assert(m_StateChangeMonitors != null);
             Debug.Assert(m_StateChangeMonitors.Length > deviceIndex);
+            Debug.Assert(m_StateChangeMonitors[deviceIndex].listeners != null);
+            
+            if (m_StateChangeMonitors == null || m_StateChangeMonitors.Length <= deviceIndex ||
```

Comment:
I think we need to reconsider this. I see the following issues (correct me if I am wrong about this....)

1) The checks in the if statement are basically the same as the Debug.Asserts (including the addition of checking listeners). It seems at first glance that the previous Debug.Asssert statements are there to catch "programmer errors" and not that its valid to call this function if there are no state change monitors or using an index out of bounds. I assume this is not happening. I am not certain that it's a good approach to just return if these conditions are not met since they likely indicate that something is logically wrong?
2) The next issue are these Debug.Assert methods in general which I believe is bad practise from an unsafe context since I believe it will crash the application/editor if not fulfilled right? Try with Debug.Assert(false) and see what happens from this method. Hence I believe we should remove the Debug.Asserts here and either throw exception if these fundamental conditions are not fulfilled or Debug.Assert them on a higher level in a non-unsafe context. It seems this method is being invoked from two places which are also in turn unsafe so it might be problematic to refactor this way even though internal....

I would propose (and let's discuss based on your opinion), that we convert those existing two Debug.Assert statements into Debug.Asserts on a higher level if possible or convert those into throwing if violated. In addition the listeners check you added here is new and it might actually be the offender here since I notice we set it to default (as mentioned in the comments of the ticket) when removing a state monitor. It might be possible that tis condition alone should be a condition for early return.

Reply:
For future reference, Håkan and I discussed this yesterday:
• Debug asserts can catch programmer errors, but in the couple of cases of crashes we've seen, it's been in release builds where they don't fire, and seemingly is caused by a timing issue so the problem doesn't happen in development builds. So, assert to catch obvious mistakes so people can improve their code, but be extra safe too, just in case.
• We believe the issue with Asserts in unsafe is specifically from the unit test setup, where we can use alternative assert functionality - standard Debug.Assert is fine to use in `unsafe` "normal" code.